92 core files

candyhazlett opened the pull request:
@amclark42, this branch now has all of the updates I've made to the core file, collection, project and user models. This is the version of the app that I run locally. I made a separate markdown file, [LOCAL_DEPLOY_GUIDE](https://github.com/NEU-DSG/tapas_rails/blob/92-core-files/LOCAL_DEPLOY_GUIDE.md), to walkthrough getting it set up in your environment. 

Fingers-crossed this works as expected, but if not feel free to comment here or Slack DM. We can merge this to development once it's confirmed the app is run-able on other computers besides my own at our next meeting.

candyhazlett committed:
removes the institution entity (and related files) as institution information will now be user-provided as a string of text, renames communities as projects, modifies views to remove references to both institutions and communities.

candyhazlett committed:
adds migrations to rename communities projects and removes account type and institution-as-entity related components from app

candyhazlett committed:
adds modifications to view templates related to removal of account type, refactoring institution input, and renaming communities -> projects

candyhazlett committed:
updates gitignore file to include tapas query syntax file and a markdown file outlining new db structure

candyhazlett committed:
adds popper js library to asset pipeline for bootstrap 5 bump

candyhazlett committed:
this file was changed when a library, likely bootstrap, was updated, but as it was done in the background, I'm not sure. committing it anyhow with a comment explaining the origin of the modification.

candyhazlett committed:
uncomments css @import statements for the core_files, projects (formerly communities), jquery-ui, and jquery-ui/resizable libraries, adding them to the asset pipeline

candyhazlett committed:
this adds a modification to the main application template in an initial effort to generate the site in xml. this is not complete.

candyhazlett committed:
updates parameters to replace ':avatar' with 'image_file[:file]' as all image files are now called image_files independent of the model to which the file is associated.

candyhazlett committed:
adds this document for reference of previous solr-af setup for storing and accessing data in the app

candyhazlett committed:
adds modifications or removes test files related to community renaming, institution entity removal, or references to AF or the previous version of the xml db connection tool

candyhazlett committed:
renames community -> project in user access definitions

candyhazlett committed:
modifies the collection class and related files to add the image_file association, rename community to project, remove some unused methods, adds a helper method for the project_collection join table and its class file,

candyhazlett committed:
modifies default encoding value to use utf8mb4. this change occurred as a result of enabling the ActiveStorage module for file attachments

candyhazlett committed:
changes directory name for files related to xml db processes

candyhazlett committed:
renames xml db service files, TapasXQService

candyhazlett committed:
removes bootstrap-sass in favor of bootstrap to upgrade from version 3 -> 5, removes sass-rails in favor of sassc-rails

candyhazlett committed:
comments out content security policy entirely to allow for testing

candyhazlett committed:
comments out command to copy blacklight config as it was causing errors. removes reference to the :did attribute since it's not part of the new version, changes Exist module calls to TapasXq. adds instructions for modifying the upsert action to use the updated TapasXq api

candyhazlett committed:
replaces thumbnail has_many association with image_files has_many through polymorphic association, removes unused solr indexing methods, removes community method and puts its logic into the project method, adds the create_view_package_methods method back after deletion by myself (accidentally) or the previous maintainers

candyhazlett committed:
changes activestorage service value to :local, and updates the url for the xml db api

candyhazlett committed:
adds console output and initial logic for attaching image files to user

candyhazlett committed:
updates method call for user image file to display in user registration edit view

candyhazlett committed:
new polymorphic model to store all image files related to user, project, collection, or core files

candyhazlett committed:
adds new name for image files

candyhazlett committed:
renaming references to community as project instead

candyhazlett committed:
updates url for xml db api

candyhazlett committed:
just renamed community models

candyhazlett committed:
Revert "adds this document for reference of previous solr-af setup for storing and accessing data in the app"

This reverts commit 5f136c291fb58f5b063fd0c998345de319e98d10.

candyhazlett committed:
reviewed together to explain rationale for changes, also found typos and logic errors that were corrected during review.

candyhazlett committed:
just makes association between the (then community) project and collection tables

candyhazlett committed:
removes collections_core_files join table in separate migration

candyhazlett committed:
auto-generated comments

candyhazlett committed:
puts path to data directory with sql import file to gitignore, and updates associations on collection class file and thumbnail helper method

candyhazlett committed:
updates name of image file class in helper method

candyhazlett committed:
removes belongs_to association reference as core_files_user is join table, each table has an habtm association with the other

candyhazlett committed:
removes belongs_to association reference as core_files_user is join table, each table has an habtm association with the other defined in the migration

candyhazlett committed:
removes belongs_to association reference as core_files_user is join table, each table has an habtm association with the other defined in the migration

candyhazlett committed:
adds todo list to gitignore, and env config template, removes institutions and reference to communities from routes, adds a todo to the collections show page for creating a form and instruction text telling user steps to delete a collection and decide where collection core files should go, adds todo for adding user instructions to project show view, updates url from exist to tapas-xq in the test and staging yml files for xml db connection, adds image_url to image_files model, datestamps to project_collections model, creates a new project_core_files join model, adds helper methods to image_file class

candyhazlett committed:
adds migrations for project_core_file, collection_core_file migrations to create new tables, and a migration for adding an image_url attribute/column to image_file model/table, adds helper method to process_image_data and uncomments file_format method to display an error when user attempts to upload an unsupported file type to image_file class

candyhazlett committed:
adds todo to collection class to create logic for returning a list of collections that a user can add core_files to before deleting a collection, adds a todo to the destroy action on the collections controller to first display instructions to user for them to move the collection core files to a new collection and confirm before delete, replaces references to community with project in the collection upsert action, recreates community upsert action as project upsert action, deletes community upsert action, updates core_file, reading_interface,  upsert action to replace Exist references with tapas-xq

candyhazlett committed:
adds todos to create methods on core file class to ensure deletion of associated join table records on a core file deletion, removes reference to datastream_id from Download path concern as that attribute is a legacy of activefedora, adds a separate task for attaching image files to dummy data records in dummy_data_generator.rake, updates mods_xml action to use the tapas-xq module, adds task to delete all records from solr index, replaces communitiespage reference with projectspage in coffee file,

candyhazlett committed:
adds todo to pages controller for refactoring request parsing logic to use table id instead of Drupal-associated id, updates image file association on project class and modifies existing tod related to user prompt on deletion, adds requisite frozen_string_literal value to project_member class file,

candyhazlett committed:
removes commented out code from new sign up form, removes institutions class reference from registrations controller, removes confirmable method from devise library reference in user class file, updates image file association, removes institutions, and renames community to project, also adds a helper method to quickly access project member role from user object, changes community to project on users_controller.rb

candyhazlett committed:
removes commented out code from new sign up form, removes institutions class reference from registrations controller, removes confirmable method from devise library reference in user class file, updates image file association, removes institutions, and renames community to project, also adds a helper method to quickly access project member role from user object, changes community to project on users_controller.rb

candyhazlett committed:
swaps out string values for environment variables in rake task to generate test data for debugging, adds a new markdown file to walk through using the refactored version of the app locally

amclark42 commented:
[My previous comment](https://github.com/NEU-DSG/tapas_rails/issues/92#issuecomment-2484006688) was accidentally left in the paired issue rather than this thread, oops. Tl;dr: I had a rough time installing the dependencies and in particular the `mysql2` gem. 

I've now successfully installed `mysql2`. I needed to provide the paths to both MySQL and OpenSSL in installing `mysql2`. This worked:

```shell
gem install mysql2 -- --with-mysql-dir=$(brew --prefix mysql) --with-openssl-dir=$(brew --prefix openssl@3)
bundle install
```

I was also able to configure Solr. (I'm not sure what schema.xml file I should be using, so I copied an older one I had from the Hydra days.)

More will have to wait until tomorrow, but I'm optimistic this will work. Whew.

amclark42 commented:
Alas, running `rails db:create` fails with `dyld[4236]: missing symbol called`. [There are two possible culprits](https://www.rubyonmac.dev/how-to-install-therubyracer-on-m1-m2-apple-silicon-mac): `mysql2`, or (most likely) `therubyracer`. My Macbook's Apple chip doesn't play well with this, it seems.

Luke from Archimedes uncommented the `therubyracer` line in the Gemfile, commit 432f1c7. That makes me think we probably don't need it.

...Okay, I commented `therubyracer` out of the Gemfile, ran `bundle clean --force`, and tried creating the database again. I now get a new error: `NameError: uninitialized constant IMAGE_BASE_URL`.

Progress? This feels like an unset environment variable or something. I'll try diving deeper next week.

amclark42 committed:
Comment out "therubyracer" gem again.

Prevents error "dyld[4236]: missing symbol called" when trying to
run Rails on a Macbook with an Apple M1 chip.

amclark42 committed:
Provide strings to `ENV.fetch()`.

Fixes "NameError: uninitialized constant IMAGE_BASE_URL" in the
dummy_data_generator.rake when I tried to run `rails db:create`.

amclark42 commented:
The error `NameError: uninitialized constant IMAGE_BASE_URL` was occurring because `ENV.fetch()` in dummy_data_generator.rake was attempting to use the _variable_ `IMAGE_BASE_URL` instead of the _string_ `"IMAGE_BASE_URL"`. I've added a commit fixing the bug.

With this change, and a `.env` in the `tapas_rails` directory, I was able to run `rails db:create` and `rails db:migrate`.

`rails dummy_data_generator:run_all` fails on adding images:
```
Adding image files...
true
Image file attached for User 1
rails aborted!
fatal: No live threads left. Deadlock?
4 threads, 4 sleeps current:0x000000011015ebc0 main thread:0x0000000120605a40
* #<Thread:0x0000000120863ac8 sleep_forever>
   rb_thread_t:0x0000000120605a40 native:0x00000001f49fb840 int:0
   /Users/aclark/.rvm/rubies/ruby-2.7.7/lib/ruby/2.7.0/monitor.rb:108:in `sleep'
   /Users/aclark/.rvm/rubies/ruby-2.7.7/lib/ruby/2.7.0/monitor.rb:108:in `wait'
```

I assume this is because I used `IMAGE_BASE_URL='https://example.org'` in my `.env` file — the task tried to put the images there and it didn't work. Today I'm going to do a little debugging and research to see if I can get the task to complete.

amclark42 committed:
Only attach images if `IMAGE_BASE_URL` is set

amclark42 commented:
H'okay, I've hit my limit for today. With commit [8abea85](https://github.com/NEU-DSG/tapas_rails/pull/93/commits/8abea85aaa349262f048ce7524daace7fe2f11e6), the dummy data task now works for me _as long as_ I don't have `IMAGE_BASE_URL` set in my `.env` file.

Reading the code for `dummy_data_generator:attach_image_files`, I figured out that `IMAGE_BASE_URL` is a misnomer — it's actually the path to an image file that gets attached to each user, collection, and project. In theory, I could use a simple filepath like `/Users/aclark/Pictures/nu-husky.png` as my `IMAGE_BASE_URL` and the task would just use that.

In practice, I can _almost_ get this working:

```ruby
def record_image(record, image_name=nil)
    url = "#{IMAGE_BASE_URL}"
    image_data = URI.open(url)
    image_name ||= "#{record.class}_#{record.__id__}"
    image_mimetype = image_data.is_a?(File) ? 'image/png' : image_data&.content_type
    depositor_id = record.is_a?(User) ? record.id : record.depositor_id

    record_assoc_image_file = ImageFile.create(
      title: image_name,
      depositor_id: depositor_id,
      imageable_type: record.class.name,
      imageable_id: record.id,
      file_format: image_mimetype,
      image_url: url
    )

    record_assoc_image_file.save

    record.image_file.attach(
      io: image_data,
      filename: record_assoc_image_file.title,
      content_type: image_mimetype
    )

    #puts record.image_file.analyze

    puts "Image file attached for #{record.class} #{record.id}" if record.image_file.attached?
  end
```

Because I'm using a `File` rather than an `IO`, I needed to provide a MIME type manually (there had been no HTTPS request with a Content Type header to grab). This led to several new errors:

- `MiniMagick::Error: You must have ImageMagick or GraphicsMagick installed` — The most straightforward message. I installed ImageMagick with Homebrew and this message went away.
- `fatal: No live threads left. Deadlock?` — The trace suggests that the problem is in `record_image()`, but I'm not sure how. Commenting out `puts record.image_file.analyze` made the problem move down to `if record.image_file.attached?`, and I lack the knowledge to understand what this means.
- `can't alloc thread (ThreadError)` — This seems related to the error above — a new thread can't be allocated, presumably because Rails reached its limit and there are still threads waiting on something.
  - `/Users/aclark/.rvm/gems/ruby-2.7.7/gems/activesupport-5.2.6/lib/active_support/dependencies.rb:511:in 'load_missing_constant': Unable to autoload constant Devise::Mailer, expected /Users/aclark/.rvm/gems/ruby-2.7.7/gems/devise-4.9.3/app/mailers/devise/mailer.rb to define it (LoadError)` — This message only appears alongside the `alloc` message. I'm guessing Rails is trying to send an email to an administrator, but something's misconfigured.

Setting the attached images aside, I checked Solr for TAPAS records. Solr is reporting 0 documents for `tapas_core` — something's gone wrong here too, I guess. I _can_ start `rails server` and see dummy collections at localhost:3000. Oddly, I have the option to edit and delete some of those collections, despite not being logged in. Clicking into anything to do with collections results in errors.

I think this might be as far as I can get on my own. Since my edits to `dummy_data_generator.rake` didn't work, I'm not going to commit them. Instead, I'll update the deploy guide with the information I've gathered here, and plan to start reviewing the actual code changes tomorrow.

amclark42 committed:
Update local deploy guide.

I built out the instructions for installing the system
dependencies, so the novice isn't left on their own to figure stuff
out. Also added some explanations and info about dependencies that
were accidentally left out.

Since Rails doesn't fully work for me, this might need some more
editing. :/

amclark42 commented on app/actions/tapas_xq/concerns/mods_xml.rb:
It looks like `app/actions/tapas_xq/concerns/mods_xml.rb` is a duplicate of [app/actions/exist/concerns/mods.rb → app/actions/tapas_xq/concerns/tapas_xq.rb](https://github.com/NEU-DSG/tapas_rails/pull/93/files#diff-336d9d71a1625da21ad46ec6ee17d3318136ba5f7ad4983880e068ef2cb17ff3). Can we delete this one in favor of the composite?

candyhazlett replied:
Nice catch! I made issue #94 to make sure I don't forget to resolve this.

candyhazlett replied:
This is assigned to issue #94

amclark42 commented on app/actions/tapas_xq/get_mods.rb:
```diff
@@ -1,10 +1,10 @@
-    include Exist::Concerns::Helpers
-    include Exist::Concerns::Mods
+    include TapasXq::Concerns::Helpers
+    include TapasXq::Concerns::ModsXML
```
Okay, this line suggests that `mods_xml.rb` might still have a use. In which case, `tapas_xq.rb` may not need the MODS module, since it duplicates the code in `ModsXML`? Some detangling or cleanup will be needed here, though I realize that TAPAS-xq isn't the focus of this PR.

candyhazlett replied:
This is the type of matter I would have brought to you eventually anyhow, so I'm happy to get clarity on it beforehand even if it's not directly related to the PR.

candyhazlett replied:
issue #95

amclark42 commented on app/actions/upsert_project.rb:
```diff
@@ -0,0 +1,51 @@
+  private
+
+    def update_metadata!(project)
+      # project.mods.title = params[:title] if params.has_key? :title
+      project.DC.title = params[:title] if params.has_key? :title
+      # project.mods.abstract = params[:description] if params.has_key? :description
+      project.DC.description = params[:description] if params.has_key? :description
+      project.match_dc_to_mods
```
Another thing which shouldn't hold up the PR! This code is mapping TAPAS Rails metadata to `DC`, or [Dublin Core](https://www.dublincore.org/specifications/dublin-core/). Dublin Core is a simple, flexible, and powerful metadata schema that we needed to use for Fedora. Then, the code appears to map the Dublin Core fields to their corresponding fields in [MODS](https://www.loc.gov/standards/mods/), a different metadata schema that TAPAS-xq uses and Rails serves out.

All this to say, there's a lot of hoop-jumping here that we can streamline later!

candyhazlett replied:
issue #96

amclark42 commented on app/assets/stylesheets/application.css.scss:
```diff
@@ -22,13 +22,13 @@
 
-// @import 'jquery-ui';
+ @import 'jquery-ui';
 
-// @import 'jquery-ui/resizable';
+ @import 'jquery-ui/resizable';
```
We won't need jQuery or jQuery UI! Newer versions of Bootstrap don't rely on it, and I'm not planning to use it for the interface.

candyhazlett replied:
I created issue #97 to review the gem requirements and confirm whether or not jquery-ui is used elsewhere and thus, has to be imported.

amclark42 commented on app/controllers/core_files_controller.rb:
```diff
@@ -93,7 +93,12 @@ def update
 
   def view_package_html
-    @core_file = CoreFile.find_by_did(params[:did])
+    # the :did attribute has been removed since it's not part of the new version of the app
+    # but it may need to be added again to store legacy did's? not sure yet
+    # decide what to do with did values that are associated with core files in the current prod database as those files are imported into the new database
+    core_file_id = params[:id] || params[:did]
+
+    @core_file = CoreFile.find_by_id(core_file_id)
```
The original pattern `CoreFile.find_by_did(params[:did])` is handled several different ways in this file, which makes me think we should figure out a future-proofed way of handling Drupal identifiers *and* Rails IDs sooner rather than later. This way we can implement it and use it consistently.

I wrote a whole thing musing about how to solve this (I like the way of handling the problem demonstrated in this code selection), then I realized that the code we're replacing here is ultimately doing something different than what we need. In the old code, Drupal was _handing Rails_ a request for some record via its Drupal ID. The TAPAS user and Drupal itself didn't know or care what the Rails ID was. And Rails could rely on _only_ receiving a Drupal-style ID.

Our problem now is that we need URLs previously generated by Drupal to be routed to their Rails equivalent. We won't necessarily know whether an ID in a URL is Drupal-style or Rails-style (though we may be able to make an educated guess). Also, the logic could turn out to be quite complex, because Drupal accepts multiple kinds of URL "identifier"s _for the same resource_.

I'm putting an item on my To Do list to research the Drupal-style URLs we'll have to support. In the meantime, it might be a good idea to add a method to the controller with all the logic that needs to be repeated, e.g.:
```ruby
    core_file_id = params[:id] || params[:did]

    CoreFile.find_by_id(core_file_id)
```

...this way we only need to update code in that method when we figure out how to obtain the right CoreFile.

candyhazlett replied:
This seems like an important piece to resolve either before merging this PR, or immediately after as a separate issue once this PR is merged.

candyhazlett replied:
I created issue #99 for this. My initial thought is that the import process we plan to use for getting the production app db data into the v2 app db will also guide this task of refactoring methods related to identifying core files for requests to the controller overall. 

i.e., we can store both the v2 app db generated IDs and URLs for core files and the production db DIDs and URLs for core files by blending the v2 app core file db table structure with the production core file db table structure (dropping whatever fields we know we won't need to persist to v2) when creating the sql script to generate the db dump file to import. 

once the blended table exists in the v2 app db, it will be simpler to refactor this logic, though the logic itself will be fairly complex as you stated.

amclark42 commented on app/controllers/core_files_controller.rb:
```diff
@@ -149,15 +154,20 @@ def api_show
-      if CoreFile.exists_by_did?(params[:did])
-        core_file = CoreFile.find_by_did(params[:did])
+      if CoreFile.exists?(params[:id])
+        core_file = CoreFile.find_by_id(params[:id])
```
Since this line's part of an upsert, we have more control over the URL, and can probably depend on having an actual ID. We can just use `CoreFile.find(params[:id])`.

amclark42 commented on app/models/core_file.rb:
```diff
@@ -98,8 +111,28 @@ def retroactively_set_status!
 
-  # Check to see if this is an ography-type upload or a tei file type
-  # upload
+  def create_view_package_methods
+    array = available_view_packages_machine
+
+    array.each do |method_name|
+      string_name = method_name
+      method_name = method_name.to_sym
+      CoreFile.send :define_method, method_name do |arg = :models|
+        if arg.blank?
+          arg = :models
+        end
+        tg = self.content_objects(:raw).find do |x|
+          #change this to whatever field name replaced "active_fedora_model_ssi"
+          x["active_fedora_model_ssi"] == "HTMLFile" &&
+            x["html_type_ssi"] == string_name
+        end
+
+        load_specified_type(tg, arg)
+      end
+    end
+  end
```
I'm really curious about this method, but I can't really tell what the code's doing. Some comments would help.

amclark42 commented on app/models/image_file.rb:
```diff
@@ -0,0 +1,20 @@
+class ImageFile < ActiveRecord::Base
+  # associations
+  belongs_to :imageable, polymorphic: true
+  belongs_to :depositor, class_name: "User"
```
I could see this getting us into trouble. If a user uploads an image to be used for a project, and then deletes their account, shouldn't the image remain as long as the project remains available? It might make more sense to have images "belong" to the object they're attached to.

amclark42 commented on app/models/user.rb:
```diff
@@ -21,25 +22,27 @@ class User < ActiveRecord::Base
 
-  ACCOUNT_TYPES = %w[free teic teic_inst]
 
   def api_key=(api_key)
     @api_key = Digest::SHA512.hexdigest api_key
     self.encrypted_api_key = @api_key
   end
```
This can be removed. The referenced API is Wild Apricot, aka Pluot, which managed TEI Consortium membership. Since we're not matching TEI memberships with our users, we don't need to query their service.

amclark42 commented on app/views/core_files/_reading_interface.html.haml:
```diff
@@ -1,21 +1,22 @@
-          = reader_view.content.content.html_safe.force_encoding('utf-8')
-        - else
-          = "This view package could not be rendered"
+  - unless @view_packages.nil?
```
Noticing a couple of lines like this, future-proofing in case we don't have any view packages. Testing to ensure there's something to display is great. Good work!

amclark42 commented on app/views/layouts/application.html.erb:
XHTML! My heart is aflutter. Thank you!

amclark42 commented on app/views/users/index.html.haml:
```diff
@@ -36,6 +36,6 @@
           - unless user.discarded_at.nil?
             = user.discarded_at
-            = link_to "Really delete", user_path(user.id), class: ['btn', 'btn-default', 'btn-sm', 'btn-danger'], method: :delete, data: { confirm: "Are you sure you want to delete #{user.name} for real? This record will no longer exist." }
+            = link_to "Really delete", user_path(user.id), class: ['btn', 'btn-default', 'btn-sm', 'btn-danger'], method: :delete, data: { confirm: "Are you sure you want to delete #{user.name} for real? This record will no longer tapas_xq." }
```
Whoops, this needs to be changed back.

amclark42 commented on config/initializers/content_security_policy.rb:
We probably don't want to implement this until we've got something live on the web. Is it possible to only initialize this if we're in a production environment?

amclark42 commented on config/routes.rb:
```diff
@@ -36,36 +36,38 @@
-  # delete "communities/:did" => "communities#destroy"
+  # Projects, formerly 'Communities'
+  resources :projects
+  get 'projects/id' => 'projects#show'
+  post "projects/id" => "projects#upsert"
+  get 'projects/id/edit' => 'projects#edit'
```
It's occurring to me that we should probably discuss URL design with Scott and Julia. Using Rail's IDs is a good, simple, unique method, but they aren't human-readable like existing TAPAS users might be used to seeing. (I kind of don't want to enforce uniqueness across another field by implementing human-readable names. But from a usability perspective, readable URLs are way preferable. urghhhh 😣)

amclark42 replied:
Is it okay to leave out the colon in `project/id` and the like? I'm not sure if what I'm seeing is an intentional, systematic change or a find-replace mistake.

amclark42 commented on config/settings/development.yml:
```diff
@@ -1,5 +1,5 @@
   username: 'tapas'
   password: 'tapas'
-  url: 'http://localhost:8868/exist/apps/tapas-xq'
+  url: 'http://localhost:8868/tapas_xq/apps/tapas-xq'
```
The TAPAS-xq URLs should be `http://localhost:8080/tapas_xq` for development and testing. In production and staging, we'll probably need an environment variable for the AWS instance's URL.

I can change this myself if it's easier on you. I'm holding back on additional updates while I'm reviewing your code.

amclark42 commented on lib/tasks/dummy_data_generator.rake:
```diff
@@ -1,104 +1,199 @@
     )
+
+    puts "Admin user #{user.id} has been created." unless user.nil?
+  end
```
Ah, I didn't look at this part of the dummy data generator. If I execute `:run_all` twice, will the script create a new admin user with the dummy email address and password? Does User check for uniqueness of the email address before creating a new account?

(I'm capable of looking this up or running a test to find out, but I only have a little more time before the Thanksgiving holiday. I'll leave this for our future selves to investigate. Sorry, y'all, but I know you've got this!)

amclark42 commented on lib/tasks/dummy_data_generator.rake:
```diff
@@ -1,104 +1,199 @@
+      project = Project.create(title: Faker::Company.bs,
+                     description: Faker::Lorem.paragraph,
+                     depositor_id: User.all.sample.id,
+                     is_public: 'false'
```
Fascinated by using a string for the value of `is_public`. Can we use Boolean values? Ruby has a reserved keyword for `true` or `false`, yeah?

In XQuery and the other XML technologies, using a string `"false"` would be OK (not overtly wrong) but probably not _the right_ way to go. I'm curious how this reads in Ruby/Rails.

amclark42 commented on lib/tasks/dummy_data_generator.rake:
```diff
@@ -1,104 +1,199 @@
+  task :generate_project_members => :environment do
+    Project.all.each do |project|
+      5.times do
+        user_id = User.all.where(admin_at: nil).sample.id
```
Can we future-proof this code (and script) so we don't touch any projects that already existed before running this task? I'm thinking it might be useful to be able to spin up a new project and 5 test users programmatically, without worrying about affecting any "real" data.

That might not be what this rake task is intended to do, I suppose. It'd help to have a comment at the top of the file describing what this file can do, what its limitations are, and when to run it.

amclark42 commented on lib/tasks/dummy_data_generator.rake:
```diff
@@ -1,104 +1,199 @@
   desc 'create public collections'
   task :generate_public_collections => :environment do
-    125.times do
+    75.times do
```
It might also be useful to be able to set the number of X records to be created.

amclark42 commented on spec/actions/exist/delete_record_spec.rb:
```diff
@@ -22,18 +22,18 @@
   end
 
-  it 'returns a 500 for dids that are not indexed in exist' do
+  it 'returns a 500 for dids that are not indexed in tapas_xq' do
```
Another small thing to change back

amclark42 commented on spec/controllers/core_files_controller_spec.rb:
```diff
@@ -202,7 +202,7 @@ def test_file(fname)
     after(:each) { ActiveFedora::Base.delete_all }
 
-    it 'raises a 404 for dids that do not exist' do
+    it 'raises a 404 for dids that do not tapas_xq' do
```
Another small tweak

candyhazlett replied:
Gah! I thought I got all these.

amclark42 commented on spec/controllers/core_files_controller_spec.rb:
```diff
@@ -265,7 +265,7 @@ def test_file(fname)
 
       # Ensure support file content has been attached
-      expect(core.thumbnail).to be_instance_of ImageThumbnailFile
+      expect(core.thumbnail).to be_instance_of Thumbnail
```
Is this still right? Are we using Thumbnail for images attached to CoreFile?

candyhazlett replied:
Good catch! No. All image files are referred to as ImageFile, independent of the associated object.

amclark42 commented on spec/models/core_file_spec.rb:
```diff
@@ -107,7 +107,7 @@
     end
 
-    it "should not have method if the view_package doesn't exist" do
+    it "should not have method if the view_package doesn't tapas_xq" do
```
Another small tweak

amclark42 commented on spec/validators/core_file_validator_spec.rb:
```diff
@@ -93,7 +93,7 @@ def params
     it 'raises an error when collection_dids references nonexistent collections' do 
       validate(params.merge(collection_dids: ['111-111']))
-      it_raises_a_single_error('collections that do not exist')
+      it_raises_a_single_error('collections that do not tapas_xq')
```
Last one!

amclark42 commented on app/models/concerns/download_path.rb:
```diff
@@ -3,8 +3,9 @@ def download_path(dsid='content')
                         :action => :show, 
-                        :id => pid, 
-                        :datastream_id => dsid, 
+                        :id => id,
+                        # TODO: determine what should replace datastream_id
+                        # :datastream_id => dsid,
```
`datastream_id` was a Samvera Hydra thing. I found [this old code snippet](https://github.com/samvera/hydra-head/blob/v9.5.0/hydra-core/app/controllers/concerns/hydra/controller/download_behavior.rb#L54-L69) which says that `datastream_id` is deprecated and people should use `file` (a filename?) instead.

candyhazlett replied:
Good context to have. Yes, filename should be used. If I remember correctly, I left this here because I wasn't sure what should replace `datastream_id`. Also, is it possible for a file to only have a datastream_id and not have a filename? Seems unlikely, but it won't hurt to write code to handle for that if it is possible.

candyhazlett replied:
issue #100

amclark42 reviewed (approved):
This is a lot of good work, Candace, thank you! There are a few small things we'll need to tweak now or at some point in the future, but nothing that needs to hold up a merge.

Besides looking through my comments, you may want to check the commits I added to the PR, and make sure you can still run Rails without `therubyracer` gem.

candyhazlett committed:
removes redundant solr index-related class method from core_file and to better make use of existing solr_helpers logic. adds a text edit to LOCAL_DEPLOY_GUIDE.md.

candyhazlett committed:
Merge remote-tracking branch 'origin/92-core-files' into 92-core-files

candyhazlett committed:
refactors callback methods on collection, core_file, and project to use a different callback that will add records to solr index and update existing records in solr index

candyhazlett committed:
removes unneeded storage files added inadvertently in last commit

candyhazlett committed:
refactors associations for simplicity, modifies html templates to reflect class structure, adds some whitespace for formatting, removes some legacy code helper methods

candyhazlett committed:
for core files, refactors associations for simplicity, removes unused callback methods resulting from refactor of associations, simplifies helper methods, adds new helper methods

candyhazlett committed:
for core files, refactors associations for simplicity, removes unused callback methods resulting from refactor of associations, simplifies helper methods, adds new helper methods, adds tei_author and tei_contributor attributes, modifies controller to accept user input for these values, updates core_file views, adds migrations for join tables between core file and both collections and projects, cleans up unneeded comments, and removes user_core_file join association along with user role as it relates to specific core files as that was done under a misunderstanding of the concepts for authors and contributors

candyhazlett committed:
removes paid_at and account_type attributes from user, updates concept of a user role to be tied with projects

candyhazlett committed:
removes requirement for deleted concern, updates labels and logic on collection show view

candyhazlett committed:
removes comments and print statement from solr helpers, adds updated schema file, changes avatar and thumbnail references to image_file, removes callback to delete solr doc before deleting db record, updates association to user for image_file, and modifies spec file to use updated author contributor attribute names

candyhazlett committed:
adds logic for image file attachment and updates core_file creation with tei_author names in data generator task, updates my_tapas html template, simplifies project associations, callbacks, roles for project members, updates project params,

candyhazlett committed:
commented out references to canonical object until core_file to TEI file association is done, clarifies canonical_object, removes user email from public account profile, corrects user profile route, adds title of core file to show view along with edit button for admin users, updates logic for displaying project members on show view, formats users_controller methods
